// fastify-decorators/decorators/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var service = require('./service.cjs');
var inject = require('./inject.cjs');
var initializer = require('./initializer.cjs');
var hook = require('./hook.cjs');
var errorHandler = require('./error-handler.cjs');
var controller = require('./controller.cjs');
var requestHandlers = require('./request-handlers.cjs');

/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */

exports.Service = service.Service;
exports.Inject = inject.Inject;
exports.Initializer = initializer.Initializer;
exports.readyMap = initializer.readyMap;
exports.Hook = hook.Hook;
exports.ErrorHandler = errorHandler.ErrorHandler;
exports.Controller = controller.Controller;
exports.ALL = requestHandlers.ALL;
exports.DELETE = requestHandlers.DELETE;
exports.GET = requestHandlers.GET;
exports.HEAD = requestHandlers.HEAD;
exports.OPTIONS = requestHandlers.OPTIONS;
exports.PATCH = requestHandlers.PATCH;
exports.POST = requestHandlers.POST;
exports.PUT = requestHandlers.PUT;
