// fastify-decorators/decorators/request-handlers.js
/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */
import { requestDecoratorsFactory } from './helpers/request-decorators.factory.js';
export function ALL(config, options) {
    return requestDecoratorsFactory('all')(config, options);
}
export function GET(config, options) {
    return requestDecoratorsFactory('get')(config, options);
}
export function POST(config, options) {
    return requestDecoratorsFactory('post')(config, options);
}
export function PUT(config, options) {
    return requestDecoratorsFactory('put')(config, options);
}
export function PATCH(config, options) {
    return requestDecoratorsFactory('patch')(config, options);
}
export function DELETE(config, options) {
    return requestDecoratorsFactory('delete')(config, options);
}
export function OPTIONS(config, options) {
    return requestDecoratorsFactory('options')(config, options);
}
export function HEAD(config, options) {
    return requestDecoratorsFactory('head')(config, options);
}
