// fastify-decorators/decorators/request-handlers.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var requestDecorators_factory = require('./helpers/request-decorators.factory.cjs');

/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */
function ALL(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('all')(config, options);
}
function GET(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('get')(config, options);
}
function POST(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('post')(config, options);
}
function PUT(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('put')(config, options);
}
function PATCH(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('patch')(config, options);
}
function DELETE(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('delete')(config, options);
}
function OPTIONS(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('options')(config, options);
}
function HEAD(config, options) {
    return requestDecorators_factory.requestDecoratorsFactory('head')(config, options);
}

exports.ALL = ALL;
exports.DELETE = DELETE;
exports.GET = GET;
exports.HEAD = HEAD;
exports.OPTIONS = OPTIONS;
exports.PATCH = PATCH;
exports.POST = POST;
exports.PUT = PUT;
