// fastify-decorators/decorators/helpers/class-properties.js
/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */
import { ERROR_HANDLERS, HANDLERS, HOOKS, SERVICE_INJECTION } from '../../symbols/index.js';
import { Container } from './container.js';
export function ensureHandlers(target) {
    if (!Object.prototype.hasOwnProperty.call(target, HANDLERS)) {
        Reflect.defineProperty(target, HANDLERS, {
            value: new Container(target[HANDLERS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
export function hasHandlers(target) {
    return HANDLERS in target;
}
export function ensureErrorHandlers(target) {
    if (!Object.prototype.hasOwnProperty.call(target, ERROR_HANDLERS)) {
        Reflect.defineProperty(target, ERROR_HANDLERS, {
            value: new Container(target[ERROR_HANDLERS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
export function hasErrorHandlers(target) {
    return ERROR_HANDLERS in target;
}
export function ensureHooks(target) {
    if (!Object.prototype.hasOwnProperty.call(target, HOOKS)) {
        Reflect.defineProperty(target, HOOKS, {
            value: new Container(target[HOOKS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
export function hasHooks(target) {
    return HOOKS in target;
}
export function ensureServiceInjection(target) {
    if (!Object.prototype.hasOwnProperty.call(target, SERVICE_INJECTION)) {
        Reflect.defineProperty(target, SERVICE_INJECTION, {
            value: new Container(target[SERVICE_INJECTION]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
export function hasServiceInjection(val) {
    return ['function', 'object'].includes(typeof val) && val !== null && SERVICE_INJECTION in val;
}
