// fastify-decorators/decorators/helpers/class-properties.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../../symbols/index.cjs');
var container = require('./container.cjs');

/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */
function ensureHandlers(target) {
    if (!Object.prototype.hasOwnProperty.call(target, index.HANDLERS)) {
        Reflect.defineProperty(target, index.HANDLERS, {
            value: new container.Container(target[index.HANDLERS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
function hasHandlers(target) {
    return index.HANDLERS in target;
}
function ensureErrorHandlers(target) {
    if (!Object.prototype.hasOwnProperty.call(target, index.ERROR_HANDLERS)) {
        Reflect.defineProperty(target, index.ERROR_HANDLERS, {
            value: new container.Container(target[index.ERROR_HANDLERS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
function hasErrorHandlers(target) {
    return index.ERROR_HANDLERS in target;
}
function ensureHooks(target) {
    if (!Object.prototype.hasOwnProperty.call(target, index.HOOKS)) {
        Reflect.defineProperty(target, index.HOOKS, {
            value: new container.Container(target[index.HOOKS]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
function hasHooks(target) {
    return index.HOOKS in target;
}
function ensureServiceInjection(target) {
    if (!Object.prototype.hasOwnProperty.call(target, index.SERVICE_INJECTION)) {
        Reflect.defineProperty(target, index.SERVICE_INJECTION, {
            value: new container.Container(target[index.SERVICE_INJECTION]),
            enumerable: false,
            configurable: false,
            writable: false,
        });
    }
}
function hasServiceInjection(val) {
    return ['function', 'object'].includes(typeof val) && val !== null && index.SERVICE_INJECTION in val;
}

exports.ensureErrorHandlers = ensureErrorHandlers;
exports.ensureHandlers = ensureHandlers;
exports.ensureHooks = ensureHooks;
exports.ensureServiceInjection = ensureServiceInjection;
exports.hasErrorHandlers = hasErrorHandlers;
exports.hasHandlers = hasHandlers;
exports.hasHooks = hasHooks;
exports.hasServiceInjection = hasServiceInjection;
