// fastify-decorators/decorators/helpers/patch-methods.js
import { ERROR_HANDLERS, FASTIFY_REPLY, FASTIFY_REQUEST, HANDLERS, HOOKS, SERVICE_INJECTION } from '../../symbols/index.js';
import { hasErrorHandlers, hasHandlers, hasHooks, hasServiceInjection } from './class-properties.js';
export function patchMethods(constructor) {
    patchHandlers(constructor);
    patchErrorsHandlers(constructor);
    patchHooks(constructor);
}
function patchHandlers(constructor) {
    if (hasHandlers(constructor))
        for (const it of constructor[HANDLERS])
            patchMethod(constructor, it.handlerMethod);
}
function patchErrorsHandlers(constructor) {
    if (hasErrorHandlers(constructor))
        for (const it of constructor[ERROR_HANDLERS])
            patchMethod(constructor, it.handlerName);
}
function patchHooks(constructor) {
    if (hasHooks(constructor))
        for (const it of constructor[HOOKS])
            patchMethod(constructor, it.handlerName);
}
function patchMethod(constructor, methodName) {
    const _original = constructor.prototype[methodName];
    constructor.prototype[methodName] = function methodProxy(request, reply, ...rest) {
        return _original.call(createProxy(this, request, reply), request, reply, ...rest);
    };
}
function createProxy(target, request, reply) {
    return new Proxy(target, {
        get(target, p) {
            const value = target[p];
            if (value === FASTIFY_REQUEST)
                return request;
            if (value === FASTIFY_REPLY)
                return reply;
            if (hasServiceInjection(value))
                return createProxy(value, request, reply);
            return value;
        },
        /**
         * Avoid creating proxies over proxies by telling that already proxied class does not have any service injection
         */
        has(target, p) {
            if (p === SERVICE_INJECTION)
                return false;
            return p in target;
        },
    });
}
