// fastify-decorators/decorators/helpers/patch-methods.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../../symbols/index.cjs');
var classProperties = require('./class-properties.cjs');

function patchMethods(constructor) {
    patchHandlers(constructor);
    patchErrorsHandlers(constructor);
    patchHooks(constructor);
}
function patchHandlers(constructor) {
    if (classProperties.hasHandlers(constructor))
        for (const it of constructor[index.HANDLERS])
            patchMethod(constructor, it.handlerMethod);
}
function patchErrorsHandlers(constructor) {
    if (classProperties.hasErrorHandlers(constructor))
        for (const it of constructor[index.ERROR_HANDLERS])
            patchMethod(constructor, it.handlerName);
}
function patchHooks(constructor) {
    if (classProperties.hasHooks(constructor))
        for (const it of constructor[index.HOOKS])
            patchMethod(constructor, it.handlerName);
}
function patchMethod(constructor, methodName) {
    const _original = constructor.prototype[methodName];
    constructor.prototype[methodName] = function methodProxy(request, reply, ...rest) {
        return _original.call(createProxy(this, request, reply), request, reply, ...rest);
    };
}
function createProxy(target, request, reply) {
    return new Proxy(target, {
        get(target, p) {
            const value = target[p];
            if (value === index.FASTIFY_REQUEST)
                return request;
            if (value === index.FASTIFY_REPLY)
                return reply;
            if (classProperties.hasServiceInjection(value))
                return createProxy(value, request, reply);
            return value;
        },
        /**
         * Avoid creating proxies over proxies by telling that already proxied class does not have any service injection
         */
        has(target, p) {
            if (p === index.SERVICE_INJECTION)
                return false;
            return p in target;
        },
    });
}

exports.patchMethods = patchMethods;
