// fastify-decorators/symbols/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * @license
 * Copyright Andrey Chalkin <L2jLiga@gmail.com> (https://github.com/L2jLiga). All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/L2jLiga/fastify-decorators/blob/master/LICENSE
 */
const CREATOR = Symbol.for('fastify-decorators.creator');
const INITIALIZER = Symbol.for('fastify-decorators.initializer');
const DESTRUCTOR = Symbol.for('fastify-decorators.destructor');
const ERROR_HANDLERS = Symbol.for('fastify-decorators.error-handlers');
const HANDLERS = Symbol.for('fastify-decorators.handlers');
const HOOKS = Symbol.for('fastify-decorators.hooks');
const SERVICE_INJECTION = Symbol.for('fastify-decorators.service-injection');
const CLASS_LOADER = Symbol.for('fastify-decorators.class-loader');
const FASTIFY_REQUEST = Symbol.for('fastify-decorators.fastify-request');
const FASTIFY_REPLY = Symbol.for('fastify-decorators.fastify-reply');
const FastifyInstanceToken = Symbol.for('fastify-decorators.fastify-instance');
const FastifyRequestToken = Symbol.for('fastify-decorators.token.fastify-request');
const FastifyReplyToken = Symbol.for('fastify-decorators.token.fastify-instance');
const INITIALIZED = Symbol.for('fastify-decorators.initializer-called');

exports.CLASS_LOADER = CLASS_LOADER;
exports.CREATOR = CREATOR;
exports.DESTRUCTOR = DESTRUCTOR;
exports.ERROR_HANDLERS = ERROR_HANDLERS;
exports.FASTIFY_REPLY = FASTIFY_REPLY;
exports.FASTIFY_REQUEST = FASTIFY_REQUEST;
exports.FastifyInstanceToken = FastifyInstanceToken;
exports.FastifyReplyToken = FastifyReplyToken;
exports.FastifyRequestToken = FastifyRequestToken;
exports.HANDLERS = HANDLERS;
exports.HOOKS = HOOKS;
exports.INITIALIZED = INITIALIZED;
exports.INITIALIZER = INITIALIZER;
exports.SERVICE_INJECTION = SERVICE_INJECTION;
